### ISSUE TITLE
Use binutils-gdb from tricore-gcc-toolchain 11.3 repository as base

### ISSUE THREAD

**Issue by NoMore201** [NoMore201/tricore-gdb#4]
https://github.com/EEESlab/tricore-gcc-toolchain-11.3.0
This is a complete free and open source implementation of tricore-elf-gcc based on GCC 11.3. The compiler works good and supports many architecture version and devices.

Check if it is possible to use this version of tricore-binutils-gdb as base for gdb, or directly include a working gdb inside the toolchain

**Comment by NoMore201** [comment]
Work in progress in other repository: https://github.com/NoMore201/tricore-binutils-gdb/tree/feature/gdb-support

**Comment by go2sh** [comment]
Hey,
if you are interested: https://github.com/go2sh/tricore-binutils-gdb.

I have developed this from scratch and I'm testing it against an openocd port: https://github.com/go2sh/aurix-openocd

**Comment by NoMore201** [comment]
Impressive work! I did a quick test with your implementation using mainline QEMU, I get the following error:

```
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /home/iezzid1/Workspace/aurix-cmake-code-sample/test.elf...
(gdb) set architecture
TriCore:V1.1    TriCore:V1.3    TriCore:V1_6    TriCore:V1_6_2  auto
TriCore:V1.2    TriCore:V1.3.1  TriCore:V1_6_1  TriCore:V1_8
(gdb) set architecture TriCore:V1_6_1
The target architecture is set to "TriCore:V1_6_1".
(gdb) target remote localhost:1234
Remote debugging using localhost:1234
Remote 'g' packet reply is too long (expected 140 bytes, got 176 bytes): 0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800b00002000008000000000000000000000000000000000000000000000000000000000
```

Launched QEMU with the following command:

```
./qemu-system-tricore -M KIT_AURIX_TC277_TRB -kernel ../../aurix-cmake-code-sample/test.elf -s -S -display none
```

If you agree, I would like to use your implementation and contribute to it. I would also like to check if issue #5 is solved with your implementation.

On a side note, how does openocd work? Do I just need a USB JTAG adapter and run the openocd server? Do I need a DAP based debugger for AURIX?

**Comment by NoMore201** [comment]
About the issue I'm facing, I think this is caused by the fact that QEMU expects more than 35 registers from the gdb, and their index may be different from the enum you implemented:

https://github.com/qemu/qemu/blob/7872e5fdf38ac0d8d0083aabb98d67da1f530ef4/target/tricore/gdbstub.c#L108-L120

I think the total number of system register should be 43 as implemented in the patch here:

https://github.com/NoMore201/tricore-gdb/blob/e5df2263a28486c584dc090bd1cc1b5faa739a72/tricore-gdb/0006-Rebase-tricore-implementation-from-10.0.50-to-11.2.patch#L14598-L14645

**Comment by go2sh** [comment]
Openocd uses the TAS Server from Infineon and the TAS server handles the on board debugger or miniwiggler. It supports every development board from IFX. Openocd handles the flashing, reset and breakpoint/steping interaction and has a gdbserver integrated (like QEMU).

I can add those registers to gdb, I did not need them yet.  I need to update openocd and gdb together.

**Comment by NoMore201** [comment]
Cool, I'll later check openocd with miniwiggler, would be great to directly flash/debug source code from IDE without having to use expensive (and usually buggy) debuggers.

I did some further testing of your GDB after adding the missing registers, it seems to work fine but problem #5 is still there. I'll try to work on an implementation of stack frame unwinding using CSAs.

**Comment by go2sh** [comment]
Yes, that true with the stack frames. It always uses the dwarf frames which do not work properly. This needs to be investigated a bit deeper.  The code for the unwinding itself is not to difficult but the problem is how to get gdb to use the unwinding function and not the dwarf info.

**Comment by go2sh** [comment]
I found the source file in quemu. Strange ordering, I was sorting them by address...
https://gitlab.com/qemu-project/qemu/-/blob/master/target/tricore/gdbstub.c

**Comment by go2sh** [comment]
I fixed the register you mentioned.
